src/ray/common/scheduling/scheduling_alg.h

```diff
@@ -0,0 +1,147 @@
+#ifndef RAY_COMMON_SCHEDULING_SCHEDULING_H
+#define RAY_COMMON_SCHEDULING_SCHEDULING_H
+
+#include <iostream>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+
+/// Maximum number of predefined resources.
+#define NUM_PREDIFINED_RESOURCES 4
+/// List of predefined resources.
+/// The number of elements in enum should be equal to NUM_PREDIFINED_RESOURCES.
+enum PredefinedResources {CPU = 0, MEM = 1, GPU = 2, TPU = 3};
+
+/// Total and available capacity of a resource instance.
+struct ResourceCapacity {
+  int64_t total;
+  int64_t available;
+};
+
+/// Resource request.
+struct ResourceReq {
+  /// Amount of resource being requested.
+  int64_t demand;
+  /// Specify whether the request is soft or hard.
+  /// If hard, the entire request is denied if the demand exceeds the resource
+  /// availability. Otherwise, the request can be still be granted.
```

Comment:
Should leave a comment about the semantics of soft requests (i.e., preference given to nodes with fewer soft violations)